src/triton_fil/triton_tensor.cuh

```diff
@@ -128,23 +164,33 @@ class TritonTensor {
       // originally allocated it non-const then cast to const for consistency
       // with Triton-provided buffers. Since this is happening in the
       // destructor, removing const at this point should be safe.
-      cudaFree(reinterpret_cast<void*>(const_cast<non_const_T*>(buffer)));
+      if (target_memory_ == TRITONSERVER_MEMORY_GPU) {
+        cudaFree(reinterpret_cast<void*>(const_cast<non_const_T*>(buffer)));
+      } else {
+        std::free(reinterpret_cast<void*>(const_cast<non_const_T*>(buffer)));
+      }
     }
   }
 
   TritonTensor(const TritonTensor<T>& other)
       : name_{other.name_}, shape_{other.shape_}, dtype_{other.dtype_},
-        size_bytes_{other.size_bytes_}, is_owner_{true}, stream_{other.stream_},
-        buffer{[&] {
-          non_const_T* ptr_d =
-              allocate_device_memory<non_const_T>(other.size());
-          try {
-            raft::copy(ptr_d, other.buffer, other.size(), stream_);
-          }
-          catch (const raft::cuda_error& err) {
-            throw TritonException(
-                TRITONSERVER_errorcode_enum::TRITONSERVER_ERROR_INTERNAL,
-                err.what());
+        target_memory_{other.target_memory_}, size_bytes_{other.size_bytes_},
+        is_owner_{true}, stream_{other.stream_}, buffer{[&] {
+          non_const_T* ptr_d;
+          if (other.target_memory_ == TRITONSERVER_MEMORY_GPU) {
```

Comment:
I see these kinds of copies being repeated a few times. Can this be pulled out in a utility function with the try-catch and letting the function maybe even handle the host copy part?

Reply:
I created a new utility function `memory_copy`.